Epics build on rocky9

bhill-slac opened the pull request:
I got epics-build working on rocky9, specifically psbuild-rocky9, with this branch.
Primarily, I do two things.
1.  Added test for existence of package under git@github.com:slac-epics as part of the default location check.   This used to just look around our AFS git area if package wasn't found in modulelist.txt.   Now it also looks on github.    This allows epics-build to work for modules already imported to github:slac-epics w/o needing a decision on how modulelist.txt will be moved and/or updated.
2. Modified the code that determines if a package is an IOC, extension, or module, to make an EPICS module the default if unable to determine from the URL and packageName.

Note:  I'm seeing issues with github API errors that keep epics-build from building more than around 10 packages.   When I first saw this I couldn't even get through one package, but I worked to minimize the number of remote github operations and now can get around 10 or so.    I looked into some info on github.com re rate limiting and was able to fetch rate limit data.   I didn't see any obvious problem w/ the rate limit data.  i.e.  No counts showed me exceeding limits, but somehow I still get seemingly random failures.

bhill-slac committed:
Updated git_utils to also ck for pkg under DEF_GITHUB_REPOS

bhill-slac committed:
Modified Releaser to default to package type module if not an IOC or extension.

Hack to get around not having substring 'module' in the url.

bhill-slac committed:
Fixed errors in gitRepo.CheckoutRelease.

Check for tag already checked out was incorrect leading to unneeded refetch.
Also fixed places where code assumed the current URL was the same as
remote origin's url.  That fails for repos originally cloned from AFS repos
that are now hosted on github.

bhill-slac committed:
Reduced github api accesses to avoid limits.

Was getting many github api errors as code was generating tons
of queries for github paths that don't exist.

bhill-slac committed:
Diagnostics, particulary for determinePathToGitRepo.

Can be quite helpful when trying to debug epics-build when
AFS may or may not be available, and most but not all EPICS repos
have been moved to github.com:slac-epics

JJL772 commented on git_utils.py:
```diff
@@ -411,6 +415,19 @@ def determinePathToGitRepo( packagePath, verbose = False ):
+        print( "determinePathToGitRepo: gitRoot is %s" % gitRoot if gitRoot else None )
+
+    # Check github DEF_GITHUB_REPOS, just add the package name
+    defRepoPath = DEF_GITHUB_REPOS + '/' + gitPackageDir
+    if 'github.com' in gitRoot:
+        # This results in an extra github API access, but is needed to test if the package is hosted on github
+        repoTags = gitGetRemoteTags( defRepoPath, verbose=verbose )
+        if len(repoTags) > 0:
+            if verbose:
+                print( "determinePathToGitRepo: github %s repo is %s" % (packagePath, defRepoPath) )
+            return defRepoPath
```
Another option here might be to check if `$GIT_SITE_TOP/path/to/repo.git/hooks/pre-receive` exits with an error not. The modules & packages that we migrated to GitHub have a pre-receive hook that prints an error message and exits return code 1. If the pre-receive hook exits with 0, then you can do the extra github.com check.

JJL772 reviewed (dismissed):
Looks good to me, besides one comment/suggestion. Thanks for fixing this stuff!

bhill-slac commented:
Looks like we can't try to execute the pre-receive hook.
It would work for the new hook script you created as it just does some echoes and an exit 1.
The default pre-receive hook still used by the repos we haven't moved to github yet
has a while loop that waits till the git connection has read all the pre-receive sets.
If we try to run it standalone it waits forever.

We may need to find another solution for this.
Ideally we should be able to test if the AFS repo is valid, but it looks like a clone or fetch operation
would succeed without running any of the hook scripts.

If you've pushed all the tags, branches, etc from the AFS git repo to github, why
not just delete the AFS git repo to avoid possible confusion over which one to use?

Note that the code below which tries to fetch the tags from a possible github repo
is only run if we are unable to access eco_modulelist/modulelist.txt or don't find
the specified package in it.

Who handles the actual merge once these pull requests have been approved?

On 3/3/2025 5:12 PM, Jeremy L. wrote:

BEWARE: This email originated outside of our organization. DO NOT CLICK links or attachments unless you recognize the sender and know the content is safe.


@JJL772 approved this pull request.

Looks good to me, besides one comment/suggestion. Thanks for fixing this stuff!

________________________________

In git_utils.py<https://github.com/slac-epics/eco_tools/pull/12#discussion_r1978446506>:

> +    defRepoPath = DEF_GITHUB_REPOS + '/' + gitPackageDir
+    if 'github.com' in gitRoot:
+        # This results in an extra github API access, but is needed to test if the package is hosted on github
+        repoTags = gitGetRemoteTags( defRepoPath, verbose=verbose )
+        if len(repoTags) > 0:
+            if verbose:
+                print( "determinePathToGitRepo: github %s repo is %s" % (packagePath, defRepoPath) )
+            return defRepoPath


Another option here might be to check if $GIT_SITE_TOP/path/to/repo.git/hooks/pre-receive exits with an error not. The modules & packages that we migrated to GitHub have a pre-receive hook that prints an error message and exits return code 1. If the pre-receive hook exits with 0, then you can do the extra github.com check.

—
Reply to this email directly, view it on GitHub<https://github.com/slac-epics/eco_tools/pull/12#pullrequestreview-2655674604>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABPDOJT2AAXL2FMAFSCOMZT2ST4XVAVCNFSM6AAAAABYID52OKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMNJVGY3TINRQGQ>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

JJL772 commented:
> If you've pushed all the tags, branches, etc from the AFS git repo to github, why
not just delete the AFS git repo to avoid possible confusion over which one to use?

I left them intact to avoid "breaking" all existing clones. Since it's been a couple of months since things have been migrated, most of the AFS repos can probably be deleted.

A few have not been migrated to GitHub yet due to invalid commits. Those will need to have their git histories rewritten. Unfortunately GitHub doesn't allow us to disable `git fsck` checks on push.

> Who handles the actual merge once these pull requests have been approved?

Either of us. I was holding off on merging because I wasn't sure if Marcio wanted to review or not. After talking to him, I'm going to go ahead with merging these.

marciodo commented:
Another ugly solution would be to check if there's a tar.gz for the Git repository in AFS:
```
└─>  ls /afs/slac.stanford.edu/g/cd/swe/git/repos/package/epics/modules/
a16vme.git                                                  ethercat_has_been_moved_to_github_see_readme.tar                  perfMeasure_has_been_moved_to_github_see_readme.tar
a16vme_has_been_moved_to_github_see_readme.tar              ethercatmc.git                                                    pgp.git
acm.git                                                     ethercatmc_has_been_moved_to_github_see_readme.tar                pgp_has_been_moved_to_github_see_readme.tar
acm_has_been_moved_to_github_see_readme.tar                 ether_ip.git                                                      plcAdmin.git
acqiris.git                                                 ether_ip_has_been_moved_to_github_see_readme.tar                  plcAdmin_has_been_moved_to_github_see_readme.tar
acqiris_has_been_moved_to_github_see_readme.tar             etherPSC.git                                                      pmac.git
ADAndor3.git                                                etherPSC_has_been_moved_to_github_see_readme.tar                  pmac_has_been_moved_to_github_see_readme.tar
ADAndor3_has_been_moved_to_github_see_readme.tar            ev2_driver.git                                                    PSCD_Camac.git
ADAndor.git                                                 ev2_driver_has_been_moved_to_github_see_readme.tar                PSCD_Camac_has_been_moved_to_github_see_readme.tar
ADAndor_has_been_moved_to_github_see_readme.tar             event2.git                                                        push2slac-epics.sh
ADAravis.git                                                event2_has_been_moved_to_github_see_readme.tar                    push2slac-epics-via-pat.sh
ADAravis_has_been_moved_to_github_see_readme.tar            event.git                                                         push2slac-epics-via-ssh.sh
ADArchon.git                                                event_has_been_moved_to_github_see_readme.tar                     pva2pva.git
ADArchon_has_been_moved_to_github_see_readme.tar            evrClient.git                                                     pva2pva_has_been_moved_to_github_see_readme.tar
ADCore.git                                                  evrClient_has_been_moved_to_github_see_readme.tar                 pvAccessCPP.git
ADCore_has_been_moved_to_github_see_readme.tar              exampleCPP.git                                                    pvAccessCPP_has_been_moved_to_github_see_readme.tar
```
Or to check the README_github file in the AFS $GIT_ROOT directory:
```
  1 spectrometer is now located at https://github.com/slac-epics/spectrometer.git
  2 pvAccessCPP is now located at https://github.com/slac-epics/pvAccessCPP.git
  3 pvDataCPP is now located at https://github.com/slac-epics/pvDataCPP.git
  4 pvDatabaseCPP is now located at https://github.com/slac-epics/pvDatabaseCPP.git
  5 pvExampleTestCPP is now located at https://github.com/slac-epics/pvExampleTestCPP.git
  6 pvCommonCPP is now located at https://github.com/slac-epics/pvCommonCPP.git
  7 pvaClientCPP is now located at https://github.com/slac-epics/pvaClientCPP.git
  8 pvaClientTestCPP is now located at https://github.com/slac-epics/pvaClientTestCPP.git
  9 pva2pva is now located at https://github.com/slac-epics/pva2pva.git
 10 pvaSrv is now located at https://github.com/slac-epics/pvaSrv.git
 11 normativeTypesCPP is now located at https://github.com/slac-epics/normativeTypesCPP.git
```

bhill-slac committed:
Add ADEURESYS to pkgNamesToMacroNames.py

bhill-slac committed:
Add --url option to epics-release

bhill-slac committed:
Prioritize github.com:/slac-epics for default origin URL

bhill-slac committed:
Fixed epics-release determination of installDir for github.com:/slac-epics repos

bhill-slac committed:
Added ck in epics-release for dirty repo.

If repo has unknown, modified, or added files
that haven't been committed, it prompts the user
re a dirty repo and shows the repo status.
Added Repo.isDirty() and Repo.ShowStatus() functions,
currently stubbed for all but gitRepo subclass.

bhill-slac commented:
I've rebased this branch to reflect other pull requests that have been merged already.
I'm considering just resubmitting this branch under a new name.
Initially I just got it working for epics-update, but as I now realize from testing and adding more
fixes for epics-release as well, the majority of these commits are needed to properly support
using github.com for our default master repo location no matter which host we build on.
Note there are several commits here that have been separately submitted as pull requests.
As those get merged I'll rebase them out of this branch.

bhill-slac commented:
Resubmitting as new pull request rebased against latest master branch.
